bin/build-mb/src/build/licenses.clj

```diff
@@ -0,0 +1,201 @@
+(ns build.licenses
+  "Functionality to take a classpath and generate a file containing all libraries and their respective licenses.
+  Intended to be called from from the command line like:
+
+  clj -X mb.license/process :classpath a-classpath :output-filename a-filename
+
+  or from clojure code like:
+  ```
+  (license/process {:classpath classpath
+                    :backfill (edn/read-string (slurp (io/resource \"overrides.edn\")))
+                    :output-filename (.getAbsolutePath output-file)
+                    :report? false})
+  ```
+
+  Allows for optional backfill, either edn or a filename that is edn. Backfill should be of the form
+
+  {\"group\" {\"artifact\" \"license text\"}
+   \"group\" {\"artifact\" {:resource \"filename-of-license\"}}
+
+  :override/group {\"group\" \"license\"
+                   \"group\" {:resource \"filename-of-license\"}}
+  }
+
+  At the moment assumes a leiningen classpath which is primarily jars and your own source paths, so only detects
+  license information from jars. In the future a strategy and heuristics could be determined for when source is
+  available from local roots and git dependencies."
+  (:require [clojure.data.xml :as xml]
+            [clojure.edn :as edn]
+            [clojure.java.io :as io]
+            [clojure.string :as str])
+  (:import (java.util.jar JarFile JarFile$JarFileEntry)))
+
+(set! *warn-on-reflection* true)
+
+(def classpath-separator (System/getProperty "path.separator"))
+
+(defn jar-file? [filename]
+  (str/ends-with? filename "jar"))
+
+(defn jar-files [jar-filename]
+  (iterator-seq (.entries (JarFile. (io/file jar-filename)))))
+
+(def pom-filter (filter (fn [^JarFile$JarFileEntry jar-entry]
+                          (str/ends-with? (.getName jar-entry) "pom.xml"))))
+
+(defn jar->pom
+  "Given a jar filename, look for an adjacent pom file otherwise look for a pom file in the jar. Return the parsed xml."
+  [jar-filename]
+  (let [adjacent-pom (io/file (str/replace jar-filename #"jar$" "pom"))]
+    (if (.exists adjacent-pom)
+      (xml/parse (io/input-stream adjacent-pom)
+                 :skip-whitespace true)
+      (when-let [jar-pom (first (into [] pom-filter (jar-files jar-filename)))]
+        (xml/parse (.getInputStream (JarFile. ^String jar-filename) jar-pom)
```

Comment:
99% sure you need to `with-open` this input stream too unless `xml/parse` closes it for you. Which would be weird. Although generally closing it twice doesn't hurt.